src/lists.rs

```diff
@@ -283,43 +290,132 @@ where
-            // Use block-style only for the last item or multiline comments.
-            let block_style = !formatting.ends_with_newline && last ||
-                comment.trim().contains('\n') ||
-                comment.trim().len() > width;
-
-            let formatted_comment = try_opt!(rewrite_comment(
+            let block_style = !formatting.ends_with_newline && last;
+            let mut formatted_comment = try_opt!(rewrite_post_comment(
+                formatting.config,
+                formatting.shape,
                 comment,
+                &cloned_items,
+                inner_item,
+                i,
+                &mut item_max_width,
+                item_last_line_width,
+                last,
                 block_style,
-                Shape::legacy(width, offset),
-                formatting.config,
             ));
 
             if !formatted_comment.starts_with('\n') {
-                result.push(' ');
+                let mut comment_alignment =
+                    post_comment_alignment(item_max_width, inner_item.len());
+                if first_line_width(&formatted_comment) + last_line_width(&result) +
+                    comment_alignment + 1 > formatting.config.max_width()
+                {
+                    item_max_width = None;
+                    formatted_comment = try_opt!(rewrite_post_comment(
+                        formatting.config,
+                        formatting.shape,
+                        comment,
+                        &cloned_items,
+                        inner_item,
+                        i,
+                        &mut item_max_width,
+                        item_last_line_width,
+                        last,
+                        block_style,
+                    ));
+                    comment_alignment = post_comment_alignment(item_max_width, inner_item.len());
+                }
+                for _ in 0..(comment_alignment + 1) {
+                    result.push(' ');
+                }
+                // An additional space for the missing trailing comma
+                if last && item_max_width.is_some() && !separate {
+                    result.push(' ');
+                }
+            }
+            if formatted_comment.contains('\n') {
+                item_max_width = None;
             }
             result.push_str(&formatted_comment);
+        } else {
+            item_max_width = None;
         }
 
         if !last && tactic == DefinitiveListTactic::Vertical && item.new_lines {
+            item_max_width = None;
             result.push('\n');
         }
     }
 
     Some(result)
 }
 
+fn rewrite_post_comment<I, T>(
+    config: &Config,
+    shape: Shape,
+    comment: &str,
+    cloned_items: &I,
+    inner_item: &str,
+    i: usize,
+    mut item_max_width: &mut Option<usize>,
+    item_last_line_width: usize,
+    last: bool,
+    block_style: bool,
+) -> Option<String>
```

Comment:
So many arguments to a function looks a little bad. I wonder if we could extract a struct that takes all or most of these and pass far fewer. I would only do this if it also improves the design in other ways too, I'm not sure that it does here.